skyportal/handlers/api/unsourced_finder.py

```diff
@@ -0,0 +1,306 @@
+                # could not handle inputs
+                return self.error('Invalid argument for `dec`')
+            obj_id = f'{ra:0.6g}{dec:+0.6g}'
+            extra_display_string = ""
+
+        output_type = self.get_query_argument('type', 'pdf')
+        if output_type not in ["png", "pdf"]:
+            return self.error(f'Invalid argument for `type`: {output_type}')
+
+        imsize = self.get_query_argument('imsize', '4.0')
+        try:
+            imsize = float(imsize)
+        except ValueError:
+            # could not handle inputs
+            return self.error('Invalid argument for `imsize`')
+
+        if imsize < 2.0 or imsize > 15.0:
+            return self.error('The value for `imsize` is outside the allowed range')
+
+        facility = self.get_query_argument('facility', 'Keck')
+        image_source = self.get_query_argument('image_source', 'ztfref')
+        use_ztfref = self.get_query_argument('use_ztfref', True)
+        if isinstance(use_ztfref, str):
+            use_ztfref = use_ztfref in ['t', 'True', 'true', 'yes', 'y']
+
+        num_offset_stars = self.get_query_argument('num_offset_stars', '3')
+        try:
+            num_offset_stars = int(num_offset_stars)
+        except ValueError:
+            # could not handle inputs
+            return self.error('Invalid argument for `num_offset_stars`')
+
+        if facility not in facility_parameters:
+            return self.error('Invalid facility')
+
+        if image_source not in source_image_parameters:
+            return self.error('Invalid source image')
+
+        radius_degrees = facility_parameters[facility]["radius_degrees"]
+        mag_limit = facility_parameters[facility]["mag_limit"]
+        min_sep_arcsec = facility_parameters[facility]["min_sep_arcsec"]
+        mag_min = facility_parameters[facility]["mag_min"]
+
+        finder = functools.partial(
+            get_finding_chart,
+            ra,
+            dec,
+            obj_id,
+            image_source=image_source,
+            output_format=output_type,
+            imsize=imsize,
+            how_many=num_offset_stars,
+            radius_degrees=radius_degrees,
+            mag_limit=mag_limit,
+            mag_min=mag_min,
+            min_sep_arcsec=min_sep_arcsec,
+            starlist_type=facility,
+            obstime=obstime,
+            use_source_pos_in_starlist=True,
+            allowed_queries=2,
+            queries_issued=0,
+            use_ztfref=use_ztfref,
+            extra_display_string=extra_display_string,
+        )
+
+        self.push_notification(
+            'Finding chart generation in progress. Download will start soon.'
+        )
+        rez = await IOLoop.current().run_in_executor(None, finder)
+
+        filename = rez["name"]
+        image = io.BytesIO(rez["data"])
+
+        # Adapted from
+        # https://bhch.github.io/posts/2017/12/serving-large-files-with-tornado-safely-without-blocking/
+        mb = 1024 * 1024 * 1
```

Comment:
Since this is now used twice, instead of duplicating the code we can add it to the base handler as `.send` / `.download` or something like that.

Reply:
Since it's a bit of work to add things into baselayer, it's fine to simply refactor into the SkyPortal handler for now.